clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java

```diff
@@ -1527,6 +1530,43 @@ public void testProtocolMetadataOrder() {
         }
     }
 
+    @Test
+    public void testThreadSafeAssignedPartitionsMetric() throws Exception {
+        // Get the assigned-partitions metric
+        final Metric metric = metrics.metric(new MetricName("assigned-partitions", "consumer" + groupId + "-coordinator-metrics",
+                "", Collections.<String, String>emptyMap()));
+
+        // Start polling the metric in the background
+        final AtomicBoolean doStop = new AtomicBoolean();
+        final AtomicReference<Exception> exceptionHolder = new AtomicReference<>();
+        Thread poller = new Thread() {
+            @Override
+            public void run() {
+                // Poll as fast as possible to reproduce ConcurrentModificationException
+                while (!doStop.get()) {
+                    try {
+                        metric.value();
+                    } catch (Exception e) {
+                        exceptionHolder.set(e);
+                        doStop.set(true);
+                    }
+                }
+            }
+        };
+        poller.start();
+
+        // Assign two partitions to trigger a metric change that can lead to ConcurrentModificationException
+        client.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
+        coordinator.ensureCoordinatorReady();
+        subscriptions.assignFromUser(new HashSet<>(Arrays.asList(t1p, t2p)));
+        // Wait just a bit for the poller to see the metric value transition from 0.0 to 2.0
+        Thread.sleep(50);
```

Comment:
Rather than wait for 50ms, this could wait for the metric value read in `poller` to change (using `TestUtils.waitForCondition` here) since nothing really happens after that? Or the thread itself could terminate after the change in value and you could use `Thread.join` here.